libevmasm/Assembly.cpp

```diff
@@ -297,6 +503,142 @@ Json::Value Assembly::assemblyJSON(std::map<std::string, unsigned> const& _sourc
+	if (_json.isMember("sourceList"))
+	{
+		solAssert(_level == 0);
+		solAssert(_sourceList.empty());
+		for (Json::Value const& sourceName: _json["sourceList"])
+		{
+			solRequire(
+				std::find(parsedSourceList.begin(), parsedSourceList.end(), sourceName.asString()) == parsedSourceList.end(),
+				AssemblyImportException,
+				"Items in 'sourceList' array are not unique."
+			);
+			parsedSourceList.emplace_back(sourceName.asString());
+		}
+	}
+
+	solRequire(_json.isMember(".code"), AssemblyImportException, "Member '.code' is missing.");
+	solRequire(_json[".code"].isArray(), AssemblyImportException, "Member '.code' is not an array.");
+	for (Json::Value const& codeItem: _json[".code"])
+		solRequire(codeItem.isObject(), AssemblyImportException, "The '.code' array contains an item that is not an object.");
+
+	result->importAssemblyItemsFromJSON(_json[".code"], _level == 0 ? parsedSourceList : _sourceList);
+
+	if (_json[".auxdata"])
+	{
+		solRequire(_json[".auxdata"].isString(), AssemblyImportException, "Optional member '.auxdata' is not a string.");
+		result->m_auxiliaryData = fromHex(_json[".auxdata"].asString());
+		solRequire(!result->m_auxiliaryData.empty(), AssemblyImportException, "Optional member '.auxdata' is not a valid hexadecimal string.");
+	}
+
+	if (_json.isMember(".data"))
+	{
+		solRequire(_json[".data"].isObject(), AssemblyImportException, "Optional member '.data' is not an object.");
+		Json::Value const& data = _json[".data"];
+		for (Json::ValueConstIterator dataIter = data.begin(); dataIter != data.end(); dataIter++)
+		{
+			solAssert(dataIter.key().isString());
+			std::string dataItemID = dataIter.key().asString();
+			Json::Value const& dataItem = data[dataItemID];
+			if (dataItem.isString())
+			{
+				solRequire(
+					dataItem.asString().empty() || !fromHex(dataItem.asString()).empty(),
+					AssemblyImportException,
+					"The value for key '" + dataItemID + "' inside '.data' is not a valid hexadecimal string."
+				);
+				result->m_data[h256(fromHex(dataItemID))] = fromHex(dataItem.asString());
+			}
+			else if (dataItem.isObject())
+			{
+				size_t index{};
+				try
+				{
+					// Using signed variant because stoul() still accepts negative numbers and
+					// just lets them wrap around.
+					int parsedDataItemID = std::stoi(dataItemID);
+					solRequire(parsedDataItemID >= 0, AssemblyImportException, "The key '" + dataItemID + "' inside '.data' is out of the supported integer range.");
+					index = static_cast<size_t>(parsedDataItemID);
+				}
+				catch (std::invalid_argument const&)
+				{
+					solThrow(AssemblyImportException, "The key '" + dataItemID + "' inside '.data' is not an integer.");
+				}
+				catch (std::out_of_range const&)
+				{
+					solThrow(AssemblyImportException, "The key '" + dataItemID + "' inside '.data' is out of the supported integer range.");
+				}
+
+				if (result->m_subs.size() <= index)
+					result->m_subs.resize(index + 1);
```

Comment:
Why are we doing this manually? To prevent `m_subs` (vector) to grow by the typical 1.5/2 factor?

Reply:
That's not even the biggest problem here. I expect that this bit will be rewritten anyway because we need to deal with this:

> - [ ] [Segfault on source indexes out of range](https://github.com/ethereum/solidity/pull/13673#discussion_r1367256424)